---
title: "RNASeq Analysis for "
author: 'Informatician: '
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# Set eval to FALSE to KNIT without running the long code blocks
<<<<<<< HEAD
eval = TRUE
=======
eval = FALSE
>>>>>>> 0c8a50a24d348bf8d0226a503faa0026246f90d6
```

```{r load libraries, echo = FALSE, message = FALSE}
library("pipelineTools")
library("biomaRt")
library("tximport")
library("DESeq2")
library("ggplot2")
library("vsn")
library("gplots")
library("calibrate")
library("genefilter")
```
 
# RNASeq Analysis for Simon Fisher  

## Results Directory
Description of the contents of each directory
### Raw reads
This directory contains the raw sequence data, in separate files and are compressed using gzip. They are in the fastq format, details of the format can be found here https://en.wikipedia.org/wiki/FASTQ_format, the quality encoding is the Sanger format (also called Illumina 1.8+ or Phred+33). These files should be stored in more than one place for data security and will be required to be publicly available for publication purposes.

### Trimmed reads
This directory contains the reads that have been trimmed for contaminating sequence adapters and low quality bases. The parameters for quality trimming are bases are trimmed if the Phred score (https://en.wikipedia.org/wiki/Phred_quality_score) falls below 15, reads are removed if the length of the reads is trimmed to less than 54 bp, if the sequence data is paired end then the pair is also removed.

### FastQC
This directory has the quality control information for both the raw reads in html format.

### FastP-QC
This directory has the Fastp QC results for the raw and the timmed reads.

### Screen
This directory contains the results files from the program fastq-screen, which aligns a subset of the reads per sample to a selection of genomes to check for cross contamination.

### MultiQC
This is an HTML formatted file, called "multiqc_report.html", that has a summary of the QC steps and the alignment rates for all of the samples.

### Hisat2 Alignments
This directory has the alignment files in sample sub directories. 

### Stringtie 
Stringtie assembles RNASeq aligned reads into potential transcripts and creates abundances for those transcripts, the results files are in sample sub directories. The results from stringtie can be used for input to DESeq2.

### DESeq2
This directory contains the analysis quality control files and the pairwise comparisons.
For analysis quality control there are two PCA plots, one on the Variance Stabilisation Transformed (VST) data and another for the log transformed data, there is also a MDS plot. These plots are use to check the variability within the groups and spot outliers.

There is also an expression plot for the gene with the largest p adjusted value, this is used as a quality control for the analysis.

There are files with the normalised counts, called gene.rld.csv and gene.vst.csv.

There are groups of three files for each of the pairwise comparisons, differentially expressed gene list, a heat map and a MA plot

#### Differentially expressed gene list
The Results for the comparisons between groups are in named folders and the contain the following:

* **comparison.genes.csv** -    
This is the main results file, it is sorted on the p adjusted value and contains the following columns:  
baseMean - mean of normalized counts for all samples  
log2FoldChange - log2 fold change (MAP): condition treated vs untreated  
lfcSE - standard error: condition treated vs untreated  
stat - Wald statistic: condition treated vs untreated  
pvalue - Wald test p-value: condition treated vs untreated  
padj - BH adjusted p-values  
external_gene_name - common name for the gene  
chromosome_name - chromosome name  
start_position - start position of the gene  
end_position - end position of the gene  
gene_biotype - biotype of the gene e.g. protein coding, long non-coding RNA etc  
Columns labelled with the replicate names - the normalised counts for the replicates in the comparisons.  

* **Heatmaps** -    
comparison.0.05.heatmap.pdf - A heatmap for the genes in the comparison with a P adjusted value (padj) value of equal to or less than 0.05.  

* **Volcano plots** -    
comparison_volcano_plot.pdf - Volcano plot for the genes in the comparison, red points have a P adjusted value of less than0.1, orange points have a log 2 fold change  greater than 1 and green points have a P adjusted value less than 0.1 and a log 2 fold change  greater than 1.   

* **MA plots** -    
comparison.genes.plotMA.pdf - The MA plot shows the log2 fold changes in the comparison over the mean of the normalised counts.  Genes with an FDR value below a threshold (here 0.1) are shown in red. 

### RNASeq Standard text
Sequencing libraries were prepared from total RNA using the Illumina TruSeq Stranded mRNA Sample Preparation Kit. Libraries were sequenced in 75 base, paired end mode on the Illumina NextSeq 500 platform.
Raw sequence reads were trimmed for contaminating sequence adapters and poor quality bases using the program FastP^1^. Bases with an average Phred score lower than 15 were trimmed. Reads that were trimmed to less than 54 bases were discarded. The quality of the reads were checked using the Fastqc program (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) before and after trimming. The sequenced reads where checked for other species contamination using the Fastq-Screen^2^ program, which selects a sample of the reads and aligns them to a selection of reference genomes that are routinely sequenced at Glasgow Polyomics.
The quality trimmed reads were aligned to the reference genome using the program Hisat2^4^ and the transcript abundance estiamted using Stringtie^5^. The quality control and alignment reports were combined for all samples and presented in one report using the program MultiQC^4^.
The differential expression for the analysis groups were assessed using the Bioconductor packages Tximport^6^ and DESeq2^7^.

### References
1. Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu; fastp: an ultra-fast all-in-one FASTQ preprocessor, Bioinformatics, Volume 34, Issue 17, 1 September 2018, Pages i884–i890, https://doi.org/10.1093/bioinformatics/bty560
2. Wingett, Steven W, and Simon Andrews. FastQ Screen: A tool for multi-genome mapping and quality control. F1000Research vol. 7 1338. 24 Aug. 2018, https://dx.doi.org/10.12688%2Ff1000research.15931.2
3. Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller. MultiQC: Summarize analysis results for multiple tools and samples in a single report Bioinformatics (2016). http://dx.doi.org/10.1093/bioinformatics/btw354
4. Pertea M, Kim D, Pertea GM, Leek JT, Salzberg SL Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown, Nature Protocols 11, 1650-1667 (2016), https://doi.org/10.1038/nprot.2016.095
5. Pertea M, Pertea GM, Antonescu CM, Chang TC, Mendell JT	& Salzberg SL. StringTie enables improved reconstruction of a transcriptome from RNA-seq reads. Nature Biotechnology 2015, https://doi.org/10.1038/nbt.3122
6. Soneson C, Love MI, Robinson MD (2015). Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences. F1000Research, 4. doi: https://doi.org/10.12688/f1000research.7563.1
7. Love, M. I., Huber, W., & Anders, S. (2014). Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology, 15(12), 550. http://doi.org/10.1186/s13059-014-0550-8

## Set up  
### Paths to software
Set the paths to the software installed on the system
```{r software paths, echo = FALSE}
fastq.screen.path <- "/software/fastq_screen_v0.13.0/fastq_screen"
fastp.path <- "/software/bin/fastp"
hisat.path <- "/software/hisat_v2-2.1.0/hisat2"
multiqc.path <- "/usr/local/bin/multiqc"
samtools.path <- "/software/samtools_v1.9/samtools"
stringtie.path <- "/software/stringtie-1.3.6/stringtie"
```
Table of paths to the required software installed on the system  

|Software|Path|
|--------|----|
|FastQ Screen|`r fastq.screen.path`|
|FastP|`r fastp.path`|
|HiSat2|`r hisat.path`|
|MultiQC|`r multiqc.path`|
|Samtools|`r samtools.path`|
|Stringtie|`r stringtie.path`|

### Software Versions
Version numbers for the software used
```{r versions, warning=FALSE, echo = FALSE}
fastqscreen.version <- run_fastq_screen(fastq_screen = fastq.screen.path, version = TRUE)
fastp.version <- run_fastp(fastp = fastp.path, version = TRUE)
hisat.version <- run_hisat2(hisat2 = hisat.path, version = TRUE)
multiqc.version <- run_multiqc(multiqc = multiqc.path, version = TRUE)
samtools.version <- run_samtools(samtools = samtools.path, version = TRUE)
stringtie.version <- run_stringtie(stringtie = stringtie.path, version = TRUE)
r.version <- getRversion()
deseq2.version <- packageDescription("DESeq2")$Version
```
|Software|Version|
|--------|-------|
|FastQ Screen|`r fastqscreen.version`|
|FastP|`r fastp.version`|
|HiSat2|`r hisat.version[1]`|
|MultiQC|`r multiqc.version`|
|Samtools|`r samtools.version[1]`|
|StringTie|`r stringtie.version`|
|R|`r r.version`|
|DESeq2|`r deseq2.version`|

### Create the results directories
```{r results directories, echo = TRUE}
# Trimmed reads directory
trimmed.reads.dir <- "trimmed_reads"
#Create the directory for the trimmed reads
dir.create(trimmed.reads.dir, showWarnings = FALSE)

# FastQScreen results directory
fastq.screen.dir <- "Screen"
# Create the directory for the fastq screen results
dir.create(fastq.screen.dir, showWarnings = FALSE)

# FastP results directory
fastp.results.dir <- "FastpQC"
# Create the directory for the FastP results
dir.create(fastp.results.dir, showWarnings = FALSE)

# Hisat2 alignment results directory
hisat2.alignments.dir <- "hisat2_alignments"
#Create the directory for the HiSat2 results
dir.create(hisat2.alignments.dir, showWarnings = FALSE)

# Stringtie results directory
stringtie.dir <- "stringtie"
#Create the directory for the HiSat2 results
dir.create(stringtie.dir, showWarnings = FALSE)
```

### Load the sequence sample locations into lists
Read in the fastq file paths to lists and then create sample names lists and trimmed read file paths.
```{r setup files, echo = TRUE}
<<<<<<< HEAD
reads.path <- "/path/to/reads"

reads.patt.1 <- "*_R1_001.fastq.gz$"
reads.patt.2 <- "*_R2_001.fastq.gz$"

mate1 <- list.files(path = reads.path, pattern = reads.patt.1, full.names = TRUE)
mate2 <- list.files(path = reads.path, pattern = reads.patt.2, full.names = TRUE)

mate1.out <- paste(trimmed.reads.dir,(list.files(path = reads.path, pattern = reads.patt.1, full.names = FALSE)), sep = "/")
mate2.out <- paste(trimmed.reads.dir,(list.files(path = reads.path, pattern = reads.patt.2, full.names = FALSE)), sep = "/")

sample.names <- unlist(lapply(strsplit(list.files(path = reads.path, pattern = reads.patt.1, full.names = FALSE),"_"), `[[`, 1))
=======
reads.path <- "/export/buzz2/gpfrawdata/nextseq01/FastQ/2019/MartinMcBride/1602_NS205_0362b_SimonFisher_20190726"

mate1 <- list.files(path = reads.path, pattern = "*_R1_001.fastq.gz$", full.names = TRUE)
mate2 <- list.files(path = reads.path, pattern = "*_R2_001.fastq.gz$", full.names = TRUE)

mate1.out <- paste(trimmed.reads.dir,(list.files(path = reads.path, pattern = "*_R1_001.fastq.gz$", full.names = FALSE)), sep = "/")
mate2.out <- paste(trimmed.reads.dir,(list.files(path = reads.path, pattern = "*_R2_001.fastq.gz$", full.names = FALSE)), sep = "/")

sample.names <- unlist(lapply(strsplit(list.files(path = reads.path, pattern = "*_R1_001.fastq.gz$", full.names = FALSE),"_"), `[[`, 1))
>>>>>>> 0c8a50a24d348bf8d0226a503faa0026246f90d6
```
### Sequence adapters
```{r sequence adapters, echo = TRUE}
adapter1 <- "AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC"
adapter2 <- "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT"
```
### Reference paths
The full paths to the reference genome indexes and annotation files.
```{r references, echo = TRUE}
# Path to the reference genome
genome <- "/export/buzz1/Genome/Homo_sapiens/Ensembl/GRCh38_release_79/Sequence/Indices/hisat2_index/GRCh38"

# Path to the gtf file
gtf <- "/export/buzz1/Genome/Homo_sapiens/Ensembl/GRCh38_release_79/Sequence/Transcriptome/WholeTranscriptome/Homo_sapiens_GRCh38_79_primary_assembly_whole_transcriptome.gff"
# Gene Ontology annotation database
GOAnnoDB <- "org.Hs.eg.db"
```

## Fastq Screen
Results are stored in the `r fastq.screen.dir` directory.
```{r fastqscreen, echo = TRUE, eval = eval}
fastq.screen.conf <- "/export/jessie3/gmh5n/PipelineTest/fastq_screen.conf"
fastq.screen.cmds <- run_fastq_screen(fq.files = mate1,
                                      out.dir = fastq.screen.dir,
                                      conf = fastq.screen.conf,
                                      top = "100000,5000",
                                      fastq_screen = fastq.screen.path)
write.table(fastq.screen.cmds,"FastqScreen_commands.sh", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## FastP
Adapter and quality trimmed reads are stored in the `r trimmed.reads.dir` directory, the QC files are stored in the `r fastp.results.dir` directory.
```{r fastp, echo = TRUE, eval = eval}
fastp.cmds <- run_fastp(mate1 = mate1,
                        mate2 = mate2,
                        mate1.out = mate1.out,
                        mate2.out = mate2.out,
                        adapter1 = adapter1,
                        adapter2 = adapter2,
                        sample.name =  sample.names,
                        out.dir = fastp.results.dir,
                        fastp = fastp.path)

write.table(fastp.cmds,"FastP_commands.sh", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## HISAT2
Align the reads to the genome, alignments stored in the `r hisat2.alignments.dir` directory.
```{r hisat2, echo = TRUE, eval = eval}
# Paired end
strandedness <- "RF"
hisat2.cmds <- run_hisat2(mate1 = mate1,
                          mate2 = mate2,
                          index = genome,
                          sample.name = sample.names,
                          strandedness = strandedness,
                          assembly = TRUE,
                          out.dir = hisat2.alignments.dir,
                          hisat2 = hisat.path)

write.table(hisat2.cmds,"HiSat2_commands.sh", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## MultiQC
Collate the QC and alignment statistics into one report using MultiQC the results can be viewed by opening the multiqc_report.html in a web browser.
```{r multiqc, echo = TRUE, eval = eval}
workingDir <- getwd()
run_multiqc(workingDir,
            multiqc = multiqc.path)
```

## Samtools
convert the sam format file to sorted bam format, results stored in the `r hisat2.alignments.dir` directory
```{r samtools, echo = TRUE, eval = eval}
sam.files <- list.files(path = hisat2.alignments.dir, pattern = "sam$", full.names = TRUE, recursive = TRUE)
bam.files <- gsub(sam.files, pattern = ".sam", replacement = ".bam")

# Sam to Bam
command <- "view"
samtools.cmds <- run_samtools(command = command,
                              input = sam.files,
                              sample.name = sample.names,
                              samtools = samtools.path)
samtools.cmds

# Sort
command <- "sort"
samtools.cmds <- run_samtools(command = command,
                              input = bam.files,
                              sample.name = sample.names,
                              samtools = samtools.path)
samtools.cmds

# Index
command <- "index"
samtools.cmds <- run_samtools(command = command,
                              input = bam.files,
                              samtools = samtools.path)
samtools.cmds
```

```{r remove files, eval = eval}
file.remove(sam.files)
file.remove(bam.files)
```
## Stringtie
Assemble transcripts for each sample, results stored in the `r stringtie.dir` directory.
```{r stringtie assemble, echo = TRUE, eval = eval}
sorted.bam.files <- list.files(path = hisat2.alignments.dir, pattern = "sorted.bam$",full.names = TRUE, recursive = TRUE)

stringtie.assemble.cmds <- run_stringtie(input = sorted.bam.files, 
                                         trimming = FALSE,
                                         strandedness = "first",
                                         reference.gtf = gtf,
                                         estimate = TRUE,
                                         out = stringtie.dir,
                                         sample.name = sample.names,
                                         stringtie = stringtie.path)

write.table(stringtie.assemble.cmds,"string_assemble_commands.sh", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

Merge transcripts from all samples, results stored in the `r stringtie.dir` directory.
```{r stringtie merge, echo = TRUE, eval = eval}
gtf.files <- list.files(path = stringtie.dir, pattern = ".gtf$",full.names = TRUE, recursive = TRUE)


gtf.list <- "gtf.list.txt"
write.table(gtf.files,gtf.list, quote = FALSE, row.names = FALSE, col.names = FALSE)

merged.gtf <- "stringtie_merged.gtf"

stringtie.merge.cmds <- run_stringtie(input = gtf.list,
                                      merge = TRUE,
                                      reference.gtf = gtf,
                                      out = merged.gtf,
                                      stringtie = stringtie.path)

write.table(stringtie.merge.cmds,"string_merge_commands.sh", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

Estimate transcript abundances and create table counts. Using the merged GTF file to estimate transcript abundances for Ballgown and DESeq2. Results stored in the `r stringtie.dir` directory.
```{r stringtie abundancess, echo = TRUE, eval = eval}
stringtie.abundance.cmds <- run_stringtie(input = sorted.bam.files,
                                           reference.gtf = merged.gtf,
                                           estimate = TRUE,
                                           ballgown = TRUE,
                                          merge = FALSE,
                                           out = stringtie.dir,
                                           sample.name = sample.names,
                                           stringtie = stringtie.path)
write.table(stringtie.abundance.cmds,"string_abundance_commands.sh", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## TXImport
Create a table mapping ENSEMBL transcript ids to ENSEMBL gene ids. Then create a summarised matrix of abundances, counts, and transcript lengths to the gene-level per sample.
```{r tximport, echo = TRUE, eval = TRUE}
files <- file.path(stringtie.dir,sample.names,"t_data.ctab")
names(files) <- paste0(sample.names)

file1 <- read.csv(files[1], sep = "\t")
trans.ids <- file1$t_name

mart<-"ensembl"
db<-"hsapiens_gene_ensembl"
filt<-"ensembl_gene_id"

# Create the biomaRt object
ensembl = useEnsembl(biomart=mart, dataset=db)
Att<-listAttributes(ensembl)
# Get all the transcript ids and corresponding gene ids from BiomaRt
att<-c("ensembl_transcript_id","ensembl_gene_id")
txTable<-getBM(attributes=att,values=trans.ids,mart=ensembl)
# Set the column naes for the transcript to gene table
colnames(txTable)<-c("tx_id","gene_id")
head(txTable)

txi <- tximport(files, type = "stringtie", tx2gene = txTable, ignoreTxVersion = TRUE, ignoreAfterBar = TRUE)

```

## DESeq2
### Read in the experimental design
The sample description file has the experimental design and is used to create a matrix of all possible comparisons. This file is re-organised so it is the same order as the counts matrix.
```{r experimental design, echo = TRUE}
sampleTable <- read.csv("SampleDescription.txt", sep="\t")
sampleTable
# Contrast to be performed all cominations generated from the SampleDescription file Condition column
conditions <- unique(sampleTable[ ,"Condition"])
comparisons <- t(combn(conditions,2))

# Order the Sample Table the same as the txi object
sampleTable <- sampleTable[match(sample.names,sampleTable$Sample), ]
```
**The Sample Bought-In-UM is in a group of one, all comparisons with this group should be treated as highly suspect, whilst the differential expression software will generate P values and adjusted P values these are not real**

### Differential expression analysis
Differential expression analysis.
```{r differential expression, echo = TRUE}
DESeq2Table <- DESeqDataSetFromTximport(txi,colData=sampleTable,design= ~ Condition)
dds <- DESeq(DESeq2Table)

# Create directory for DESeq2 data
data_dir = "DESeq2"
dir.create(data_dir, showWarnings = FALSE)
```


### Quality control plots
#### Dispersion plots
The black points are the dispersion estimates for each gene as obtained by considering the information from each gene separately. Unless there are many samples, these values fluctuate around their true values. The red trend line is fitted, which shows the dispersions’ dependence on the mean, and then shrink each gene’s estimate towards the red line to obtain the final estimates (blue points) that are then used in the hypothesis test. The blue circles above the main “cloud” of points are genes which have high gene–wise dispersion estimates which are labelled as dispersion outliers. These estimates are therefore not shrunk toward the fitted trend line. These warnings just indicate that the dispersion estimation failed for some genes.  
```{r dipersion plot, echo = TRUE}
plotDispEsts(dds,main="Gene level")
pdf(file.path(data_dir,"DispPlot.pdf",fsep = .Platform$file.sep))
plotDispEsts(dds,main="Gene level")
dev.off()
```

#### Variance stabilisation
Count data transformations by regularized logarithm and variance stabilizing transformations
```{r variance stabilisation, echo = TRUE}
rld <- rlog(dds)
rldMat <- assay(rld)
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix( sampleDists )
write.csv(rldMat, file=file.path(data_dir,"gene.rld.csv",fsep = .Platform$file.sep))
vsd <- varianceStabilizingTransformation(dds)
class(vsd)
vstMat <- assay(vsd)
write.csv(vstMat, file=file.path(data_dir,"gene.vst.csv",fsep = .Platform$file.sep))
write.csv(cor(vstMat), file=file.path(data_dir,"gene.cor.vst.csv",fsep = .Platform$file.sep))
```

#### PCA Plots
rlog transformed data
```{r rlog PCA, echo = TRUE}
pca.data <- plotPCA(rld, intgroup = c("Condition"), returnData = TRUE)
percentVar <- round(100 * attr(pca.data, "percentVar"))
ggplot(pca.data, aes(PC1, PC2, label=name, color=Condition)) + geom_point(size=5) + geom_text(aes(label=name), colour="black",size=2) + xlab(paste0("PC1: ",percentVar[1],"% variance")) + ylab(paste0("PC2: ",percentVar[2],"% variance")) + coord_fixed()
pdf(file.path(data_dir,"rlog_PCA.pdf",fsep = .Platform$file.sep))
ggplot(pca.data, aes(PC1, PC2, label=name, color=Condition)) + geom_point(size=5) + geom_text(aes(label=name), colour="black",size=2) + xlab(paste0("PC1: ",percentVar[1],"% variance")) + ylab(paste0("PC2: ",percentVar[2],"% variance")) + coord_fixed()
dev.off()
```

vsd transfromed data
```{r vsd PCA , echo = TRUE}
data <- plotPCA(vsd, intgroup = c("Condition"), returnData = TRUE)
percentVar <- round(100 * attr(data, "percentVar"))
ggplot(data,aes(PC1,PC2, label=name, color=Condition)) + geom_point(size=5) + geom_text(aes(label=name), colour="black",size=2) + xlab(paste0("PC1: ",percentVar[1],"% variance")) + ylab(paste0("PC2: ",percentVar[2],"% variance")) + coord_fixed()
pdf(file.path(data_dir,"vsd_PCA.pdf",fsep = .Platform$file.sep))
ggplot(data,aes(PC1,PC2, label=name, color=Condition)) + geom_point(size=5) + geom_text(aes(label=name), colour="black",size=2) + xlab(paste0("PC1: ",percentVar[1],"% variance")) + ylab(paste0("PC2: ",percentVar[2],"% variance")) + coord_fixed()
dev.off()
```

#### MDS plot
rlog transformed data
```{r mds plot, echo = TRUE}
mdsData <- data.frame(cmdscale(sampleDistMatrix))
mds <- cbind(mdsData, as.data.frame(colData(rld)))
ggplot(mds, aes(X1,X2, color=Condition, label=rownames(mds))) + geom_point(size=5) + coord_fixed() + geom_text(colour = "black",size=2) + coord_fixed()
pdf(file.path(data_dir,"rlog_MDS.pdf",fsep = .Platform$file.sep))
  ggplot(mds, aes(X1,X2, color=Condition, label=rownames(mds))) + geom_point(size=5) + coord_fixed() + geom_text(colour = "black",size=2) + coord_fixed()
dev.off()
```
**Samples S1-UM-K18 and S1-UM-K32 appear to be different drom the other two in their respective groups, is there any chance that these could be a battch effect?**

Normalised counts for adding to the results file
```{r normalised counts, echo = TRUE}
disp<-estimateSizeFactors(dds)
norm_counts<-counts(disp,normalized=TRUE)
```

Per-gene standard deviation (taken across samples), against the rank of the mean.
```{r sd plot, echo = TRUE}
se <- SummarizedExperiment(log2(counts(disp, normalized=TRUE) + 1), colData=colData(dds))
data<-plotPCA(DESeqTransform(se),intgroup = c("Condition"), returnData = TRUE)
ggplot(data,aes(PC1,PC2, label=name, color=Condition)) + geom_point(size=5) + geom_text(aes(label=name), colour="black",size=2) + xlab(paste0("PC1")) + ylab(paste0("PC2")) + coord_fixed()

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
pdf(file.path(data_dir,"geneSdPlot.pdf",fsep = .Platform$file.sep))
#meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1))
meanSdPlot(assay(vsd[notAllZero,]))
dev.off()
```

#### Heatmap
Plot a heatmap for the genes with a p adjusted vale of 0.1 or less
```{r heatmap, echo = TRUE}
qVal <-"0.1"
hmcols<- colorRampPalette(c("green4","green","white", "red","red4"))(256)
dds_results <- results(dds)
de.genes <- rownames(dds_results)[ which(dds_results$padj < qVal) ]
if(length(de.genes) > 1){
  nt <- normTransform(dds)
  df <- as.data.frame(colData(dds)[,c("Condition")])
  log2.norm.counts <- assay(nt)[de.genes,]
  pdfName <- paste("DEGenes",qVal,"heatmap.pdf",sep = "_")
  pdf(pdfName)
  heatmap(log2.norm.counts)
  dev.off()
    pdf(file.path(data_dir,pdfName,fsep = .Platform$file.sep))
       heatmap.2(log2.norm.counts, col=hmcols, scale="row", hclust=function(x) hclust(x, method='complete'), distfun=function(x)  as.dist(1-cor(t(x))), trace="none", density.info="none", labRow=NA, cexCol = 0.7)
  dev.off()
  heatmap.2(log2.norm.counts, col=hmcols, scale="row", hclust=function(x) hclust(x, method='complete'), distfun=function(x)  as.dist(1-cor(t(x))), trace="none", density.info="none", labRow=NA, cexCol = 0.7)
}
```

#### Annotation dabase
Download annotations from biomaRt and add to the results file.
```{r include=FALSE, echo = TRUE, eval = eval}
att<-c("ensembl_gene_id","external_gene_name","chromosome_name","start_position","end_position","gene_biotype","entrezgene_id")
annotations<-getBM(attributes=att,filter=filt,values=rownames(dds),mart=ensembl)
dim(annotations)
annotations<-annotations[!duplicated(annotations$ensembl_gene_id), ]
dim(annotations)
head(annotations)
DESeq2Features <- data.frame(ensembl_gene_id = rownames(dds))
DESeq2Features$ensembl_gene_id <- as.character(DESeq2Features$ensembl_gene_id)
rowData <- dplyr::left_join(DESeq2Features, annotations, by = att[1])
rowData <- unique(as(rowData, "DataFrame"))
mcols(rowRanges(dds)) <- cbind(mcols(rowRanges(dds)),rowData)
```

#### Pairwise differential expression
Iterate throught the pairwise comparisons write out comma seperated files, MA plots, Volcano plots and heatmaps for each comparison 
```{r echo=FALSE, eval = eval}
for(i in 1:nrow(comparisons)){
  baseCondition <- as.character(comparisons[i,1])
  testCondition <- as.character(comparisons[i,2])
  
  
  # Conrast name
  contrast_name <- paste(baseCondition,"vs",testCondition, sep = "_")
  # Create directory for DESeq2 data
  results_dir = paste(contrast_name,"results", sep = "_")
  resultsFilePath <- file.path(data_dir,results_dir, fsep = .Platform$file.sep)
  dir.create(resultsFilePath)
  
  print(paste("Comnparing: ",contrast_name))
  # Get the samples names for the contrasts
  contrasts <- c(baseCondition,testCondition)
  comparisonSamples <- subset(sampleTable,sampleTable$Condition %in% contrasts)
  comparisonSampleNames <- comparisonSamples$Sample
  print(comparisonSampleNames)
  
  # From the DESeq2: res <- results(dds, contrast=c("condition","trt","con")) So the order for the contrast must be treated over control.
  comp <- results(dds,contrast=c("Condition",testCondition,baseCondition))
  
  # Annotate results with gene names
  comp$external_gene_name <- mcols(dds)$external_gene_name
  # Annotate results with chromosome
  comp$chromosome_name <- mcols(dds)$chromosome_name
  # Annotate results with start
  comp$start_position <- mcols(dds)$start_position
  # Annotate results with end
  comp$end_position <- mcols(dds)$end_position
  # Annotate results with biotypes
  comp$gene_biotype <- mcols(dds)$gene_biotype
  # Annotate results with biotypes
  comp$entrezgene <- mcols(dds)$entrezgene_id
  
  # Command below adds the normalised counts for the output files
  # Subset the log2 normailsed counts to just the commpasison samples
  subset.norm_counts <- norm_counts[,comparisonSampleNames]
  comp_merged<-merge(comp,subset.norm_counts, by = "row.names", all = TRUE)
  # Reorder results on adjusted p-values
  comp_merged_sorted <- comp_merged[order(comp_merged$pvalue),]
  # Convert into matrix
  comp_merged_sorted_matrix<-as.matrix(comp_merged_sorted)

  # Output file name
  outfileName <- paste(contrast_name,'genes','csv',sep=".")
  # Write output file
  write.csv(comp_merged_sorted_matrix, file=file.path(resultsFilePath,outfileName,fsep = .Platform$file.sep), quote=FALSE)
  
  # MA plot
  rmafile <- paste(contrast_name,'genes','plotMA','pdf',sep=".")
  pdf(file.path(resultsFilePath,rmafile, fsep = .Platform$file.sep))
  main <- paste('Genes:',contrast_name,sep=" ")
  plotMA(comp,alpha=0.1,main=main,ylim=c(-5,5))
  dev.off()
  
  # Heatmap of p adjusted genes
  de.genes <- rownames(comp)[ which(comp$pvalue < qVal) ]
  # Check that there are DE genes
  print(length(de.genes))
  if(length(de.genes) > 1){
    nt <- normTransform(dds)
    df <- as.data.frame(colData(dds)[,c("Condition")])
    # Create a matrix of log2normalised counts for all samples
    log2.norm.counts <- assay(nt)[de.genes,comparisonSampleNames]
    # Plot the heatmap
    pdfName <- paste(contrast_name,qVal,"heatmap.pdf",sep = ".")
    heatmap_name <- paste(contrast_name,qVal,"heatmap",sep = " ")
    pdf(file.path(resultsFilePath,pdfName, fsep = .Platform$file.sep))
      heatmap.2(log2.norm.counts, col=hmcols, scale="row", hclust=function(x) hclust(x, method='complete'), distfun=function(x)  as.dist(1-cor(t(x))), trace="none", density.info="none", labRow=NA, main=heatmap_name, cexCol = 0.7)
      dev.off()
  }
  
  # Volcano Plot
  # Make a basic volcano plot
  pdfName <- paste(contrast_name,"volcano_plot.pdf",sep = "_")
  pdf(file.path(resultsFilePath,pdfName, fsep = .Platform$file.sep))
    with(comp, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-2.5,2.5),ylim=c(0,20)))
    # Add colored points: red if pvalue<0.05, orange of log2FC>1, green if both)
    with(subset(comp, pvalue<.1 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
    with(subset(comp, abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="orange"))
    with(subset(comp, pvalue<.1 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
  
    # Label points
    with(subset(comp, pvalue<.1 & abs(log2FoldChange)>1), textxy(log2FoldChange, -log10(pvalue), labs=external_gene_name, cex=.8))
  dev.off()
  
  summary(comp)
  print(paste('contrast: ',contrast_name,' DONE'))
}
```

## Session Information
```{r}
sessionInfo()
```
